src/commonMain/kotlin/exchange.dydx.abacus/state/app/helper/TriggerOrderToastGenerator.kt

```diff
@@ -0,0 +1,380 @@
+            orderPayloadStatus[placeOrderPayload.clientId.toString()] = status
+        }
+    }
+
+    private fun generateTriggerOrderToast(
+        subaccountNumber: Int,
+        responsePayload: HumanReadableTriggerOrdersPayload?
+    ): List<Toast> {
+        if (responsePayload == null || state == null) {
+            return emptyList()
+        }
+        val marketId = state?.input?.triggerOrders?.marketId ?: return emptyList()
+        val market = state?.market(marketId) ?: return emptyList()
+        val subaccountNumberString = parser.asString(subaccountNumber) ?: return emptyList()
+        val selectedSubaccount =
+            state?.account?.subaccounts?.get(subaccountNumberString) ?: return emptyList()
+        val selectedSubaccountPosition = selectedSubaccount.openPositions?.firstOrNull { position ->
+            position.id == marketId && (
+                position.side.current == PositionSide.SHORT ||
+                    position.side.current == PositionSide.LONG
+                )
+        }
+        val triggerOrders = selectedSubaccount.orders?.filter { order ->
+            order.status == OrderStatus.untriggered
+        }
+        val takeProfitOrders = triggerOrders?.filter { order ->
+            selectedSubaccountPosition?.side?.current?.let { currentSide ->
+                (order.type == OrderType.takeProfitMarket || order.type == OrderType.takeProfitLimit) &&
+                    order.side.isOppositeOf(currentSide)
+            } ?: false
+        }
+        val stopLossOrders = triggerOrders?.filter { order ->
+            selectedSubaccountPosition?.side?.current?.let { currentSide ->
+                (order.type == OrderType.stopMarket || order.type == OrderType.stopLimit) &&
+                    order.side.isOppositeOf(currentSide)
+            } ?: false
+        }
+
+        val tickSize = parser.asString(market.configs?.tickSize)
+
+        val toasts = mutableListOf<Toast>()
+        responsePayload.cancelOrderPayloads?.forEach { cancelOrderPayload ->
+            val toast = generateForCancelOrder(
+                cancelOrderPayload,
+                takeProfitOrders,
+                stopLossOrders,
+                tickSize,
+            )
+            if (toast != null) {
+                toasts.add(toast)
+            }
+        }
+        responsePayload?.placeOrderPayloads?.forEach { placeOrderPayload ->
+            val toast = generateForPlaceOrder(
+                placeOrderPayload,
+                tickSize,
+            )
+            if (toast != null) {
+                toasts.add(toast)
+            }
+        }
+
+        return toasts
+    }
+
+    private fun generateForCancelOrder(
+        cancelOrderPayload: HumanReadableCancelOrderPayload,
+        takeProfitOrders: IList<SubaccountOrder>?,
+        stopLossOrders: IList<SubaccountOrder>?,
+        tickSize: String?,
+    ): Toast? {
+        if (formatter == null || localizer == null) {
```

Comment:
Why even support passing in a nullable `formatter` or `localizer`? These should never be null right?